### ISSUE TITLE
Azure Queue stream provider should guarantees at-least-once delivery

### ISSUE THREAD

**Issue by weitaolee** [dotnet/orleans#166]
https://orleans.codeplex.com/discussions/581795

I think it should open a issue:-)
```
public async Task<IEnumerable<IBatchContainer>> GetQueueMessagesAsync()
        {
            try
            {
                var task = queue.GetQueueMessages();
                outstandingTask = task;
                CloudQueueMessage[] messages = (await task).ToArray();
                if (!messages.Any())
                    return Enumerable.Empty<IBatchContainer>();
                
                AzureQueueBatchContainer[] azureQueueMessages = messages
                    .Select(msg => AzureQueueBatchContainer.FromCloudQueueMessage(msg, lastReadMessage++)).ToArray();
                //delete messages before delivery
                outstandingTask = Task.WhenAll(messages.Select(queue.DeleteQueueMessage));
                await outstandingTask;

                return azureQueueMessages;
            }
            finally
            {
                outstandingTask = null;
            }
        }
```

**Comment by gabikliot** [comment]
This is an Azure Queue Stream provider specific issue, not a general streams issue. Will be addressed soon, as part of a bigger refactoring we are doing to IQueueAdapters.

**Comment by weitaolee** [comment]
Looking forward this bigger refactoring

**Comment by weitaolee** [comment]
@gabikliot please consider support subscribe message from  AMQP MQTT STOMP  ,not only  poll message by GetQueueMessagesAsync.   because less latency  subscribe then poll .

**Comment by gabikliot** [comment]
Yes, we will support push based queues as well.

**Comment by veikkoeeva** [comment]
Is this still active?

**Comment by gabikliot** [comment]
Yes, this is still work in progress. It's a bit not trivial with the current adapter architecture. I need to track every element in the cache when it was delivered to all interested consumers. Considering multiple possible solutions.

**Comment by gabikliot** [comment]
This has been finally fixed via #770.
